src/main.rs

```diff
@@ -1112,7 +1225,9 @@ mod tests {
             .await
             .unwrap();
 
-        let consume_err = consume_messenger_events(mock, receiver, current_peers)
```

Comment:
You might have missed this comment https://github.com/carlaKC/lndk/pull/33#discussion_r1181387247

If you have any tips there lmk

Reply:
Something along these lines would work? Ideally extended to have two different pubkeys + assert that we drain all our peers messages in one go.
```
sender.send(MessengerEvents::SendOutgoing).await.unwrap();
mock.expect_next_onion_message_for_peer().return_once(|_| Some(onion_message));

let mut sender_mock = MockSendCustomMessenger::new();
sender_mock.expect_send_custom_message().return_once(|_| Ok(SendCustomMessageResponse {  }));

sender.send(MessengerEvents::SendOutgoing).await.unwrap();
mock.expect_next_onion_message_for_peer().return_once(|_| None);
```

+ remember to bump channel buffer size for new messages.

Reply:
Could also try something fancy with [matching mock arguemnts](https://docs.rs/mockall/latest/mockall/#matching-arguments) if we extend with multiple pubkeys.

Reply:
Heh this is actually almost exactly the same code that I wrote initially. But when calling `return_once`, I would get this error: `Expectation(<anything>) called twice`. `Return_once` seems to verify that the function is called so many times, rather than controls how many times the function is called.

Same with times() I believe

Reply:
> Could also try something fancy with [matching mock arguemnts](https://docs.rs/mockall/latest/mockall/#matching-arguments) if we extend with multiple pubkeys.

Seems like matching arguments it is then:
```
let mut mock = MockOnionHandler::new();

mock.expect_next_onion_message_for_peer().withf(|actual_pk: &PublicKey| *actual_pk == pubkey()).returning(|_| Some(onion_message()));
mock.expect_next_onion_message_for_peer().returning(|_| None);

let result = mock.next_onion_message_for_peer(pubkey());
assert!(result.is_some());

let result = mock.next_onion_message_for_peer(pubkey_1());
assert!(result.is_none());
```

Annoying it doesn't have "return this once", no mock is perfect 😧

Reply:
This seems to cause the same issue with the infinite loop. Because of the while let Some(msg) loop, the first `expect_next_onion_message_for_peer` just keeps spitting out the same onion message in an infinite loop. We never reach `expect_next_onion_message_for_peer` call that returns None.

Yeah it would super nice if they have a "return this once" option

I'll read a bit more on sequence and checkpoint tonight, maybe there's something in there that can help :/

Reply:
Ah yup checkpoint seems to do the trick. When calling after the first `expect_next_onion_message_for_peer` and assertion, it clears up the infinite loop. Good to know! I'll push these changes up this evening.

Reply:
Thanks for talking through this with me. A combination of matching arguments and checkpoint got it working. :)

I tweaked `fn pubkey()` to return a random pubkey instead of hardcoding the pubkeys, so that it's easy to make more than one for the tests. LMK what you think.

Reply:
WHAT a battle :') thanks for digging in to find our saviour `checkpoint` 🙏